File: src/main/java/dev/lbuddyboy/flash/user/model/Grant.java
package dev.lbuddyboy.flash.user.model;

import dev.lbuddyboy.flash.Flash;
import dev.lbuddyboy.flash.FlashLanguage;
import dev.lbuddyboy.flash.rank.Rank;
import dev.lbuddyboy.flash.util.TimeUtils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.RequiredArgsConstructor;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.TimeZone;
import java.util.UUID;

@RequiredArgsConstructor
@Data
public class Grant {

    private final UUID uuid;
    private final UUID rank;
    private final String rankName;
    private final UUID addedBy;
    private final String addedReason;
    private final long addedAt;
    private final long duration;
    private final String[] scopes;

    private UUID removedBy;
    private long removedAt;
    private String removedFor;

    public Rank getRank() {
        return Flash.getInstance().getRankHandler().getRanks().get(this.rank);
    }

    public String getRankName() {
        return getRank() == null ? this.rankName : getRank().getName();
    }

    public boolean isRemoved() {
        return removedBy != null;
    }

    public boolean isExpired() {
        return getExpiresAt() <= 0;
    }

    public long getExpiresAt() {
        return (this.addedAt + this.duration) - System.currentTimeMillis();
    }

    public String getExpireString() {
        if (duration == Long.MAX_VALUE) return "Never";
        return TimeUtils.formatLongIntoDetailedString(getExpiresAt() / 1000);
    }

    public String getAddedAtDate() {
        SimpleDateFormat sdf = new SimpleDateFormat();
        sdf.setTimeZone(TimeZone.getTimeZone(FlashLanguage.TIMEZONE.getString()));
        return sdf.format(addedAt);
    }

    public String getRemovedAt() {
        SimpleDateFormat sdf = new SimpleDateFormat();
        sdf.setTimeZone(TimeZone.getTimeZone(FlashLanguage.TIMEZONE.getString()));
        return sdf.format(removedAt);
    }

    public static Grant defaultGrant() {
        Rank actual = Flash.getInstance().getRankHandler().getDefaultRank();
        return new Grant(UUID.randomUUID(), actual.getUuid(), actual.getName(), null, "Default Grant", System.currentTimeMillis(), Long.MAX_VALUE, new String[]{"GLOBAL"});
    }

}

File: src/main/java/dev/lbuddyboy/flash/user/model/Grant.kt
package dev.lbuddyboy.flash.user.model

import dev.lbuddyboy.flash.Flash
import dev.lbuddyboy.flash.FlashLanguage
import dev.lbuddyboy.flash.rank.Rank
import dev.lbuddyboy.flash.util.TimeUtils
import lombok.*
import java.text.SimpleDateFormat
import java.util.*

@RequiredArgsConstructor
@Data
class Grant {
    private val uuid: UUID? = null
    private val rank: UUID? = null
    private val rankName: String? = null
    private val addedBy: UUID? = null
    private val addedReason: String? = null
    private val addedAt: Long = 0
    private val duration: Long = 0
    private val scopes: Array<String>
    private val removedBy: UUID? = null
    private val removedAt: Long = 0
    private val removedFor: String? = null
    fun getRank(): Rank? {
        return Flash.instance.rankHandler.getRanks().get(rank)
            ?: return Flash.instance.rankHandler.getDefaultRank()
    }

    fun getRankName(): String? {
        return if (getRank() == null) rankName else getRank().getName()
    }

    fun isRemoved(): Boolean {
        return removedBy != null || removedAt > 0 || removedFor != null
    }

    fun isExpired(): Boolean {
        return getExpiresAt() <= 0
    }

    fun getExpiresAt(): Long {
        return addedAt + duration - System.currentTimeMillis()
    }

    fun getExpireString(): String? {
        return if (duration == Long.MAX_VALUE) "Never" else TimeUtils.formatLongIntoDetailedString(getExpiresAt() / 1000)
    }

    fun getAddedAtDate(): String {
        val sdf = SimpleDateFormat()
        sdf.timeZone = TimeZone.getTimeZone(FlashLanguage.TIMEZONE.string)
        return sdf.format(addedAt)
    }

    fun getRemovedAtDate(): String {
        val sdf = SimpleDateFormat()
        sdf.timeZone = TimeZone.getTimeZone(FlashLanguage.TIMEZONE.string)
        return sdf.format(removedAt)
    }

    companion object {
        fun defaultGrant(): Grant {
            val actual: Rank = Flash.instance.rankHandler.getDefaultRank()
            return Grant(
                UUID.randomUUID(),
                actual.getUuid(),
                actual.getName(),
                null,
                "Default Grant",
                System.currentTimeMillis(),
                Long.MAX_VALUE,
                arrayOf("GLOBAL")
            )
        }
    }
}